src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.BasicBlockBuilder.cs

```diff
@@ -13,106 +13,293 @@ internal partial class ControlFlowGraphBuilder
+                {
+                    return _predecessor1;
+                }
+                else if (_predecessor1 == null)
+                {
+                    return _predecessor2;
+                }
+                else
+                {
+                    return null;
+                }
             }
 
-            internal void RemoveStatements()
+            public void AddPredecessor(BasicBlockBuilder predecessor)
             {
-                _statementsBuilder.Clear();
+                if (_predecessors != null)
+                {
+                    Debug.Assert(_predecessor1 == null);
+                    Debug.Assert(_predecessor2 == null);
+                    _predecessors.Add(predecessor);
+                }
+                else if (_predecessor1 == predecessor)
+                {
+                    return;
+                }
+                else if (_predecessor2 == predecessor)
+                {
+                    return;
+                }
+                else if (_predecessor1 == null)
+                {
+                    _predecessor1 = predecessor;
+                }
+                else if (_predecessor2 == null)
+                {
+                    _predecessor2 = predecessor;
+                }
+                else
+                {
+                    _predecessors = PooledHashSet<BasicBlockBuilder>.GetInstance();
+                    _predecessors.Add(_predecessor1);
+                    _predecessors.Add(_predecessor2);
+                    _predecessors.Add(predecessor);
+                    Debug.Assert(_predecessors.Count == 3);
+                    _predecessor1 = null;
+                    _predecessor2 = null;
+                }
             }
 
-            internal void AddPredecessor(BasicBlockBuilder block)
+            public void RemovePredecessor(BasicBlockBuilder predecessor)
             {
-                _predecessorsBuilder.Add(block);
+                if (_predecessors != null)
+                {
+                    Debug.Assert(_predecessor1 == null);
+                    Debug.Assert(_predecessor2 == null);
+                    _predecessors.Remove(predecessor);
+                }
+                else if (_predecessor1 == predecessor)
+                {
+                    _predecessor1 = null;
+                }
+                else if (_predecessor2 == predecessor)
+                {
+                    _predecessor2 = null;
+                }
             }
 
-            internal void RemovePredecessor(BasicBlockBuilder block)
+            public void GetPredecessors(ArrayBuilder<BasicBlockBuilder> builder)
             {
-                _predecessorsBuilder.Remove(block);
+                if (_predecessors != null)
+                {
+                    Debug.Assert(_predecessor1 == null);
+                    Debug.Assert(_predecessor2 == null);
+
+                    foreach (BasicBlockBuilder predecessor in _predecessors)
```

Comment:
`builder.AddRange(_predecessors);`?

Reply:
That would go through IEnumerable and I want to avoid that

---
In reply to: [192523324](https://github.com/dotnet/roslyn/pull/27363#discussion_r192523324) [](ancestors = 192523324)

Reply:
It's a nit, but I think it uses an array copy (see [here](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Builder.cs), lines 368 and 317.

Reply:
Ah, my bad. For some reason I thought it was an ArrayBuilder. Thanks

---
In reply to: [192536266](https://github.com/dotnet/roslyn/pull/27363#discussion_r192536266) [](ancestors = 192536266,192525302)